### PR TITLE
Upgrade script Implementation

### LINKED ISSUE #769: [FEATURE]: Create an Upgrade script to handle changes  in newer versions
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Category of feature request

Reconcile

### Problem statement

A new column  `operation_name` has been added to the `main` table.  
The reconcililation process is failing to write the Data to existing `main` table because of the Schema changes mentioned above.

Please find the error below:
```
AnalysisException: A schema mismatch detected when writing to the Delta table (Table ID: ).
To enable schema migration using DataFrameWriter or DataStreamWriter, please set:
'.option("mergeSchema", "true")'.
For other operations, set the session configuration
spark.databricks.delta.schema.autoMerge.enabled to "true". See the documentation
specific to the operation for details.
```

### Proposed Solution

To address these scenarios, we need to develop an Upgrade script that runs during Installation which contains detailed instructions for upgrading from one version to an upper version. 
for example: `from 0.4.0 to 0.4.1`.

### Additional Context

Check for ways to automate the process while running reconcile.
Current changes includes: Schema and Workflow

### PR DESCRIPTION
**Pull request by vijaypavann-db**
closes #769

### CODE REVIEW

File: src/databricks/labs/remorph/upgrades/v0.4.0_add_main_table_operation_name_column.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+from databricks.labs.remorph.deployment.recon import RECON_JOB_NAME
+from databricks.labs.remorph.helpers import db_sql
+
+logger = logging.getLogger(__name__)
+
+
+def _upgrade_reconcile_metadata_main_table(installation: Installation, ws: WorkspaceClient):
+    reconcile_config = installation.load(ReconcileConfig)
+    table_identifier = f"{reconcile_config.metadata_config.catalog}.{reconcile_config.metadata_config.schema}.main"
+    sql = f"ALTER TABLE {table_identifier} ADD COLUMN operation_name  STRING AFTER report_type"
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/upgrades/v0.4.0_add_main_table_operation_name_column.py:18]: how do we tackle manual interventions or table corruption?

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 2]
```diff
@@ -421,7 +421,7 @@ bad-functions = ["map", "input"]
 # ignored-parents =
 
 # Maximum number of arguments for function / method.
-max-args = 9
+max-args = 10
```
Comment (nfx) [inline pyproject.toml:424]: if method has more than 9 arguments, it's a substandard code. refactor the methods when this rule triggers. revert the change to pylint configuration.

File: src/databricks/labs/remorph/install.py [hunk 3]
```diff
@@ -298,6 +305,12 @@ def _save_config(self, config: MorphConfig | ReconcileConfig):
         if self._prompts.confirm(f"Open config file {ws_file_url} in the browser?"):
             webbrowser.open(ws_file_url)
 
+    def _apply_upgrades(self):
+        try:
+            self._upgrades.apply(self._ws)
+        except (InvalidParameterValue, NotFound) as err:
+            logger.warning(f"Installed version is too old: {err}")
```
Comment (nfx) [inline src/databricks/labs/remorph/install.py:312]: this warning is not correct - there may be other reasons.

File: src/databricks/labs/remorph/upgrades/v0.4.0_add_main_table_operation_name_column.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    db_sql.get_sql_backend(ws).execute(sql)
+    installation.save(reconcile_config)
+
+
+def _upgrade_reconcile_workflow(installation, ws):
+    context = ApplicationContext(ws)
+    context.job_deployment.deploy_recon_job(RECON_JOB_NAME, context.recon_config)
```
Comment (nfx) [inline src/databricks/labs/remorph/upgrades/v0.4.0_add_main_table_operation_name_column.py:25]: nice.

File: src/databricks/labs/remorph/install.py [hunk 5]
```diff
@@ -43,6 +47,7 @@ def __init__(
         product_info: ProductInfo,
         resource_configurator: ResourceConfigurator,
         workspace_installation: WorkspaceInstallation,
+        upgrades: Upgrades,
```
Comment (nfx) [inline src/databricks/labs/remorph/install.py:50]: add localised pylint disable comment here, i approve it for this specific case. 

other question is why we need the `ResourceConfigurator`, but it's out of scope of this PR.

**Review by nfx — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 2 above)
Comment (vijaypavann-db) [reply pyproject.toml:424]: Done

File: src/databricks/labs/remorph/install.py (hunk 3 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/install.py:312]: Updated

File: src/databricks/labs/remorph/upgrades/v0.4.0_add_main_table_operation_name_column.py (hunk 1 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/upgrades/v0.4.0_add_main_table_operation_name_column.py]: Handled

**Comment by sundarshankar89** [comment]
@bishwajit-db  can you take a look at this.

**Review by bishwajit-db — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/upgrades/v0.4.0_add_main_table_operation_name_column.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    Add operation_name column to the main table as part of the upgrade process.
+    - Compare the current main table columns with the installed main table columns. If there is any mismatch:
+       * Verify all the current main table columns are present in the installed main table and then use CTAS to recreate the main table
+       * If any of the current main table columns are missing in the installed main table, prompt the user to recreate the main table:
+            - If the user confirms, recreate the main table using the main DDL file, else log an error message and exit
+    :param installation:
+    :param ws:
+    :param app_context:
+    """
+    reconcile_config = installation.load(ReconcileConfig)
```
[the remark this answers is not in the document]
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/upgrades/v0.4.0_add_main_table_operation_name_column.py:141]: Updated

File: src/databricks/labs/remorph/install.py [hunk 7]
```diff
@@ -75,6 +84,12 @@ def run(
 
     def configure(self, module: str | None = None) -> RemorphConfigs:
         selected_module = module or self._prompts.choice("Select a module to configure:", MODULES)
+        configs = self._configure_module(selected_module)
+        self._upload_wheel()
+        self._apply_upgrades()
```
[the remark this answers is not in the document]
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/install.py:89]: Updated

File: src/databricks/labs/remorph/upgrades/v0.4.0_add_main_table_operation_name_column.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    :param installed_main_table_columns:
+    :param current_main_table_columns:
+    :param prompts:
+    :return:
+    """
+    sql: str | None = (
+        f"CREATE OR REPLACE TABLE {table_identifier} AS SELECT {','.join(current_main_table_columns)} FROM {table_identifier}"
+    )
+
+    if False in set(column in installed_main_table_columns for column in current_main_table_columns):
```
[the remark this answers is not in the document]
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/upgrades/v0.4.0_add_main_table_operation_name_column.py]: Done

**Comment by vijaypavann-db** [comment]
Had to disable pylint for the following cases:
1. `application context` exceeding allowed public methods: 
   * `pylint: disable=too-many-public-methods`
2. upgrade script file name pattern
   * `pylint: disable=invalid-name`

**Comment by vijaypavann-db** [comment]
@nfx for your review

**Comment by vijaypavann-db** [comment]
@nfx Please check this PR

**Review by nfx — APPROVED** [review]
lgtm
